modules/xfem/src/userobjects/InterfaceMeshCut3DUserObject.C

```diff
@@ -0,0 +1,406 @@
+        }
+      }
+    }
+
+    // if two edges of an element are cut, it is considered as an element being cut
+    if (cut_edges.size() == 2)
+    {
+      elem_cut = true;
+      Xfem::CutFace mycut;
+      mycut._face_id = i;
+      mycut._face_edge.push_back(cut_edges[0]);
+      mycut._face_edge.push_back(cut_edges[1]);
+      mycut._position.push_back(cut_pos[0]);
+      mycut._position.push_back(cut_pos[1]);
+      cut_faces.push_back(mycut);
+    }
+  }
+
+  return elem_cut;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutFragmentByGeometry(
+    std::vector<std::vector<Point>> & /*frag_edges*/,
+    std::vector<Xfem::CutEdge> & /*cut_edges*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutFragmentByGeometry(
+    std::vector<std::vector<Point>> & /*frag_faces*/,
+    std::vector<Xfem::CutFace> & /*cut_faces*/) const
+{
+  mooseError("cutFragmentByGeometry not yet implemented for 3D mesh cutting");
+  return false;
+}
+
+bool
+InterfaceMeshCut3DUserObject::intersectWithEdge(const Point & p1,
+                                                const Point & p2,
+                                                const std::vector<Point> & vertices,
+                                                Point & pint) const
+{
+  bool has_intersection = false;
+
+  Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+  Point point = vertices[0];
+  Point normal = elem_plane.unit_normal(point);
+
+  std::array<Real, 3> plane_point = {{point(0), point(1), point(2)}};
+  std::array<Real, 3> planenormal = {{normal(0), normal(1), normal(2)}};
+  std::array<Real, 3> edge_point1 = {{p1(0), p1(1), p1(2)}};
+  std::array<Real, 3> edge_point2 = {{p2(0), p2(1), p2(2)}};
+  std::array<Real, 3> cut_point = {{0.0, 0.0, 0.0}};
+
+  if (Xfem::plane_normal_line_exp_int_3d(
+          &plane_point[0], &planenormal[0], &edge_point1[0], &edge_point2[0], &cut_point[0]) == 1)
+  {
+    Point temp_p(cut_point[0], cut_point[1], cut_point[2]);
+    if (isInsideCutPlane(vertices, temp_p) && isInsideEdge(p1, p2, temp_p))
+    {
+      pint = temp_p;
+      has_intersection = true;
+    }
+  }
+
+  return has_intersection;
+}
+
+bool
+InterfaceMeshCut3DUserObject::isInsideEdge(const Point & p1,
+                                           const Point & p2,
+                                           const Point & p) const
+{
+  Real dotp1 = (p1 - p) * (p2 - p1);
+  Real dotp2 = (p2 - p) * (p2 - p1);
+  return (dotp1 * dotp2 <= 0.0);
+}
+
+Real
+InterfaceMeshCut3DUserObject::getRelativePosition(const Point & p1,
+                                                  const Point & p2,
+                                                  const Point & p) const
+{
+  Real full_len = (p2 - p1).norm();
+  Real len_p1_p = (p - p1).norm();
+  return len_p1_p / full_len;
+}
+
+bool
+InterfaceMeshCut3DUserObject::isInsideCutPlane(const std::vector<Point> & vertices,
```

Comment:
Same comments above. I think this method also fits better in a utility class.

Reply:
BTW, should this method be `isInsidePlane` instead of `isInsideCutPlane`? There is nothing special here related to an XFEM cut, right?